prober/derp.go

```diff
@@ -577,6 +580,196 @@ func runDerpProbeNodePair(ctx context.Context, from, to *tailcfg.DERPNode, fromc
+	// tunneling over DERP.
+	dev, _, err := tstun.New(log.Printf, "utun9")
+	if err != nil {
+		return fmt.Errorf("failed to create TUN device: %w", err)
+	}
+	defer dev.Close()
+	mtu, err := dev.MTU()
+	if err != nil {
+		return fmt.Errorf("failed to get TUN MTU: %w", err)
+	}
+
+	r, err := router.New(log.Printf, dev, nil, nil)
+	if err != nil {
+		return fmt.Errorf("failed to create router: %w", err)
+	}
+	defer r.Close()
+
+	err = r.Set(&router.Config{
+		LocalAddrs: []netip.Prefix{netip.MustParsePrefix("172.16.0.1/30")},
+		Routes:     []netip.Prefix{netip.MustParsePrefix("172.16.0.1/30")},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to set routes: %w", err)
+	}
+
+	// Read packets from TUN device
+	tunReadC := make(chan error, 1)
+	go func() {
+		// There's not a lot of thought in the below number. 5 doesn't use a ton of memory,
+		// and a bigger number like 20 didn't seem to increase throughput.
+		numBufs := 5
+		bufs := make([][]byte, 0, numBufs)
+		sizes := make([]int, numBufs)
+		for range numBufs {
+			bufs = append(bufs, make([]byte, mtu+tstun.PacketStartOffset))
+		}
+
+		for {
+			n, err := dev.Read(bufs, sizes, tstun.PacketStartOffset)
+			if err != nil {
+				return
+			}
+
+			for i := range n {
+				buf := bufs[i][:sizes[i]+tstun.PacketStartOffset]
+				pkt := buf[tstun.PacketStartOffset:]
+				src, dst := make([]byte, 4), make([]byte, 4)
+				srcPort, dstPort := make([]byte, 2), make([]byte, 2)
+				copy(src, pkt[12:16])
+				copy(dst, pkt[16:20])
+				// TODO: the below assumes no IP options in header
+				copy(srcPort, pkt[20:22])
+				copy(dstPort, pkt[22:24])
+
+				// Swap
+				copy(pkt[12:], dst)
+				copy(pkt[16:], src)
+				copy(pkt[20:], dstPort)
+				copy(pkt[22:], srcPort)
+				if err := fromc.Send(toc.SelfPublicKey(), buf); err != nil {
+					tunReadC <- err
+					return
+				}
+			}
+		}
+	}()
+
+	// Receive the packets.
+	recvc := make(chan error, 1)
+	go func() {
+		defer close(recvc) // to break out of 'select' below.
+		for {
+			m, err := toc.Recv()
+			if err != nil {
+				recvc <- fmt.Errorf("failed to receive: %w", err)
+				return
+			}
+			switch v := m.(type) {
+			case derp.ReceivedPacket:
+				if v.Source != fromc.SelfPublicKey() {
+					recvc <- fmt.Errorf("got data packet from unexpected source, %v", v.Source)
+					return
+				}
+				pkt := v.Data
+				dev.Write([][]byte{pkt}, tstun.PacketStartOffset)
+			case derp.KeepAliveMessage:
+				// Silently ignore.
+			default:
+				log.Printf("%v: ignoring Recv frame type %T", to.Name, v)
+				// Loop.
+			}
+		}
+	}()
+
+	randData := make([]byte, 10000)
+	_, err = crand.Read(randData)
+	if err != nil {
+		return fmt.Errorf("failed to initialize random data: %w", err)
+	}
+	outHash := sha256.New()
+	inHash := sha256.New()
+	var bytesWritten atomic.Uint64
+	var bytesRead atomic.Uint64
+
+	start := time.Now()
+	defer func() { transferTime.Add(time.Since(start).Seconds()) }()
+
+	// Dial ourselves
+	conn, err := net.Dial("tcp", "172.16.0.2:5201")
```

Comment:
This is pretty confusing, and I assume it relates to needing to perform L4 NAT above. Either it needs to be documented in great detail, i.e. how does this work at a syscall and handshake layer, or we should do the simpler thing and bind a listening socket first.

This also means the mutex is guarding this socket. We probably just want to listen on `:0`, resolve the port we get, and dial that. This is also justification for listen then dial, as we don't know what to dial before binding a listener.